src/herder/HerderImpl.cpp

```diff
@@ -320,13 +323,77 @@ HerderImpl::processExternalized(uint64 slotIndex, StellarValue const& value)
     // state: apply, trigger catchup, etc
     LedgerCloseData ledgerData(static_cast<uint32_t>(slotIndex),
                                externalizedSet, value);
+
+    // Only dump the most recent externalized tx set. Ledger sequence on a
+    // written tx set shall only strictly move forward; it may have gaps with
+    // the emitted debug meta, if the network is ahead of the local node
+    // (assumption is that if the network is ahead of the local node, state can
+    // be replayed from the archives)
+    if (isLatestSlot && mApp.getConfig().METADATA_DEBUG_LEDGERS != 0)
+    {
+        writeDebugTxSet(ledgerData);
+    }
+
     mLedgerManager.valueExternalized(ledgerData);
 #ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
     // Ensure potential upgrades are handled in overlay
     maybeHandleUpgrade();
 #endif
 }
 
+void
+HerderImpl::writeDebugTxSet(LedgerCloseData const& lcd)
+{
+    ZoneScoped;
+
+    // Dump latest externalized tx set. Do as much of error-handling as possible
+    // to avoid crashing core, since this is used purely for debugging.
+    auto path =
+        metautils::getLatestTxSetFilePath(mApp.getConfig().BUCKET_DIR_PATH);
+    try
+    {
+        if (fs::mkpath(path.parent_path().string()))
+        {
+            auto timer = LogSlowExecution(
+                "write debug tx set", LogSlowExecution::Mode::AUTOMATIC_RAII,
+                "took", std::chrono::milliseconds(100));
+            // If we got here, then whatever previous tx set is saved has
+            // already been applied, and debug meta has been emitted. Therefore,
+            // it's safe to just remove it.
+            std::remove(path.c_str());
+
+            auto tmpPath = path.string() + ".tmp";
+            {
+                XDROutputFileStream stream(mApp.getClock().getIOContext(),
+                                           /*fsyncOnClose=*/false);
+                stream.open(tmpPath);
+                stream.writeOne(lcd.toXDR());
+                stream.flush();
+            }
+
+            // Rename without fsync, as the debug tx set is important for
+            // debugging core crashes rather than OS crashes.
```

Comment:
I _think_ it might as well fsync here, the directory isn't likely to be very big, but I guess it adds up?

Reply:
Done.